File: GP/src/JPCNN/ImageSignature.java
package JPCNN;


public class ImageSignature extends Thread{
	RImage ri;
	public String  gSignature;
	int start ; //position in signature array to add new partial image signature
	public int sigLength; // signature array length 
	int numberofcutsitrs;
	int Niters ;
	int vert, horz;
	

	public int[] signature; // signature of a gesture (total signature of all images)
	public ImageSignature(int frameNumber, int numberofcutsitrs, int Niters, int vert, int horz){
		this.numberofcutsitrs = numberofcutsitrs;
		this.Niters = Niters;
		this.vert = vert;
		this.horz = horz;
		sigLength = frameNumber * (Niters-numberofcutsitrs) ;
		signature = new int[sigLength];
		start = 0;
		 
		gSignature = "";
		
	}
	float[][] mat;
	public ImageSignature(int frameNumber, int numberofcutsitrs, int Niters, int vert, int horz, float[][] mat){
		this(frameNumber, numberofcutsitrs, Niters, vert, horz);
		this.mat = mat;
	}
	// input path: path to image
	public void getSignature(float[][] mat) {
		RImage stim = new RImage(mat, vert, horz);
		// stim.LoadByte(path, 18 );
		// stim.loadImage(m);// I load it in constructor
		PCNN Net = new PCNN(vert, horz);
		Net.vf = 0;
		stim.div(256);
		stim.upsideDown();
		for (int i = 0; i < Niters; i++) {
			int sum = Net.iterate(stim);// Net.FastIterate(stim);
			// cout << sum <<endl;
			if (i >= numberofcutsitrs) {
				signature[i + start - numberofcutsitrs] = sum;
				gSignature += sum + " ";
			}
		}
		start += Niters - numberofcutsitrs;
	}
	@Override
	public void run() {
		getSignature(mat);
	}
}

File: bitBucket repo/pcnn1/ImageSignature.cpp
#include "ImageSignature.h"

ImageSignature::ImageSignature(int frameNumber):testRimage(dim,dim)
{
	sigLength = frameNumber * (Niters-numberofcutsitrs) ;
	signature = new elementType[sigLength];
	start = 0;
	 
	gSignature = "";
}


ImageSignature::~ImageSignature(void)
{
	delete [] signature;
}

//input path: path to image
void ImageSignature::getSignature(Mat &m)
{
	rimage stim( dim, dim );
	//stim.LoadByte(path, 18 );
	stim.loadImage(m);//I added this insted of LoadByte
    PCNN Net ( dim, dim );
    Net.vf = 0;
    stim /= 256;
    stim.UpsideDown( );
	if(testRimage == stim){
		cout << "YES OMG!";
	}else
		cout << "NO\n";
	testRimage = stim;
    int i;
    for( i=0; i<Niters; i++ ) {
		elementType sum =Net.Iterate( stim );//Net.FastIterate(stim);
		//cout << sum <<endl;
		if(i >=  numberofcutsitrs)
		{
		signature[i + start - numberofcutsitrs]= sum;
		//signature += sum + " ";
		//cout<<signature<<endl;
		char * c =new char();
		
		gSignature += _itoa(sum, c,10); 
		//delete c;
		gSignature += " ";
		}
		//cout <<"gSignature = "<< gSignature << endl;
    }
	start += Niters - numberofcutsitrs;
}

void ImageSignature::save(){
	ofstream out("dataFile.data", ios::out | ios::app);
	if(!out) { cout << "Bad file name in ImageSignature::save()\n"; exit(0);}
	//pack signatue to buffer
	//char * buffer = new char [sizeof(elementType) * sigLength];
	//memcpy(buffer, signature,sizeof(elementType) * sigLength);
	////write buffer to file 
	//out.write(buffer, sizeof(elementType) * sigLength);
	//out << signature <<endl;
	//cout <<"signature"<< signature;
	out <<"string: "<< gSignature << " " << endl;
	out << "array: ";
	for(int i = 0; i < sigLength ;i++){
		out << signature[i] << " ";
	}
	out <<endl;
	
	out.close();
}

void ImageSignature::readSignature(){
	//ifstream in("dataFile.data", ios::in | ios::binary);
	//if(!in) { cout << "Bad file name in ImageSignature::save()\n"; exit(0);}
	//char * buffer = new char [sizeof(elementType) * sigLength];
	//in.read(buffer, sizeof(elementType) * sigLength);
	////unpack buffer to signature
	//memcpy(signature, buffer, sizeof(elementType) * sigLength);
	//in.close();
}
